tests/test_graph.py

```diff
@@ -1,24 +1,74 @@
 from pythonforandroid.build import Context
-from pythonforandroid.graph import get_recipe_order_and_bootstrap
+from pythonforandroid.graph import (
+    fix_deplist, get_recipe_order_and_bootstrap, obvious_conflict_checker,
+)
 from pythonforandroid.bootstrap import Bootstrap
+from pythonforandroid.recipe import Recipe
 from pythonforandroid.util import BuildInterruptingException
 from itertools import product
 
+import mock
 import pytest
 
-
 ctx = Context()
 
 name_sets = [['python2'],
              ['kivy']]
 bootstraps = [None,
-              Bootstrap.get_bootstrap('pygame', ctx),
               Bootstrap.get_bootstrap('sdl2', ctx)]
 valid_combinations = list(product(name_sets, bootstraps))
 valid_combinations.extend(
     [(['python3crystax'], Bootstrap.get_bootstrap('sdl2', ctx)),
-     (['kivy', 'python3crystax'], Bootstrap.get_bootstrap('sdl2', ctx))])
-invalid_combinations = [[['python2', 'python3crystax'], None]]
+     (['kivy', 'python3crystax'], Bootstrap.get_bootstrap('sdl2', ctx)),
+     (['flask'], Bootstrap.get_bootstrap('webview', ctx)),
+     (['pysdl2'], None),  # auto-detect bootstrap! important corner case
+    ]
+)
+invalid_combinations = [
+    [['python2', 'python3crystax'], None],
+    [['pysdl2', 'genericndkbuild'], None],
+]
+invalid_combinations_simple = list(invalid_combinations)
+# NOTE !! keep in mind when setting invalid_combinations_simple:
+#
+# This is used to test obvious_conflict_checker(), which only
+# catches CERTAIN conflicts:
+#
+# This must be a list of conflicts where the conflict is ONLY in
+# non-tuple/non-ambiguous dependencies, e.g.:
+#
+#     dependencies_1st = ["python2", "pillow"]
+#     dependencies_2nd = ["python3crystax", "pillow"]
+#
+# This however won't work:
+#
+#     dependencies_1st = [("python2", "python3"), "pillow"]
+#     dependencies_2nd = [("python2legacy", "python3crystax"), "pillow"]
+#
+# (This is simply because the conflict checker doesn't resolve this to
+# keep the code ismple enough)
+
+
+def get_fake_recipe(name, depends=None, conflicts=None):
+    recipe = mock.Mock()
+    recipe.name = name
+    recipe.get_opt_depends_in_list = lambda: []
+    recipe.get_dir_name = lambda: name
+    recipe.depends = list(depends or [])
+    recipe.conflicts = list(conflicts or [])
+    return recipe
+
+
+def register_fake_recipes_for_test(monkeypatch, recipe_list):
```

Comment:
Remark/Taste: Since I'm a `pytest` racist I like core modules more, I would have written this using the core `unittest.mock.patch` module:
https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch

Reply:
it was pretty much the first google result I got working :smile: I have no strong opinions on this

Reply:
I understand, leave it, there're no known official best practice that say `don't use pytest fanciness`. It's just me that like to stick to core modules as much as I can :+1: